### PR TITLE
Consume all draft tokens for speculative decoding

### PR DESCRIPTION
**Pull request by as-suvorov**
Ticket: 140110

### CODE REVIEW

File: text_generation/causal_lm/cpp/speculative_decoding_lm.cpp [hunk 1]
```diff
@@ -118,150 +225,131 @@ int main(int argc, char* argv[]) try {
...
+       tokens one by one in an autoregressive manner, while the main model validates these
+       predictions and corrects them if necessary. We go through each predicted token, and
+       if a difference is detected between the draft and main model, we stop and keep the
+       last token predicted by the main model. Then the draft model gets the latest main
+       prediction and again tries to predict the next K tokens, repeating the cycle.
+
+       This approach reduces the need for multiple infer requests to the main model,
+       enhancing performance. For instance, in more predictable parts of text generation,
+       the draft model can, in best-case scenarios, generate the next K tokens that exactly
+       match the target. In tha caste the are validated in a single inference request to
```
Comment (Wovchena) [inline text_generation/causal_lm/cpp/speculative_decoding_lm.cpp:288]:
```suggestion
       match the target. In that case they are validated in a single inference call to
```
"inference request" collides with ov::InferRequest

**Review by Wovchena — CHANGES_REQUESTED** [review]
I believe your code comments are pretty clear!

File: text_generation/causal_lm/cpp/speculative_decoding_lm.cpp (hunk 1 above)
Comment (as-suvorov) [reply text_generation/causal_lm/cpp/speculative_decoding_lm.cpp]: applied

File: text_generation/causal_lm/cpp/speculative_decoding_lm.cpp [hunk 2]
```diff
@@ -118,150 +225,131 @@ int main(int argc, char* argv[]) try {
...
+       the main model (which is bigger, more accurate but slower) instead of running K
```
[the remark this answers is not in the document]
Comment (as-suvorov) [reply text_generation/causal_lm/cpp/speculative_decoding_lm.cpp]: moved to main_model assignment

File: text_generation/causal_lm/cpp/speculative_decoding_lm.cpp [hunk 3]
```diff
@@ -108,6 +105,116 @@ void update_kv_cache(ov::InferRequest request, uint64_t seq_len_axis, uint64_t n
...
+        auto sequence_logits = logits.data<float>() + (tokens_size - 1) * vocab_size;
+
+        draft_model_seq_length += tokens_size;
+
+        return std::max_element(sequence_logits, sequence_logits + vocab_size) - sequence_logits;
+    }
+
+    std::vector<int64_t> generate_candidates(const int64_t token) {
+        int64_t out_token = token;
+
```
[the remark this answers is not in the document]
Comment (as-suvorov) [reply text_generation/causal_lm/cpp/speculative_decoding_lm.cpp]: applied

File: text_generation/causal_lm/cpp/speculative_decoding_lm.cpp [hunk 4]
```diff
@@ -108,6 +105,116 @@ void update_kv_cache(ov::InferRequest request, uint64_t seq_len_axis, uint64_t n
...
+        std::vector<int64_t> candidates;
+
+        // limit candidates size by num_pred_tokens or by max_seq_length
+        size_t candidates_to_generate = std::min(num_pred_tokens, max_seq_length - draft_model_seq_length - 1);
+
+        candidates.reserve(candidates_to_generate);
+
+        // generate cadidates
+        for (size_t i = 0; i < candidates_to_generate; i++) {
+            // if out_of_kv_cache_token present - prepend it to out_token in order to collect kv cache for it
```
[the remark this answers is not in the document]
Comment (as-suvorov) [reply text_generation/causal_lm/cpp/speculative_decoding_lm.cpp]: applied

**Review by Wovchena — APPROVED** [review]

**Review by pavel-esir — APPROVED** [review]

**Pull request merged**